rust/theoros/src/handlers/websocket/subscribe_to_calldata.rs

```diff
@@ -0,0 +1,354 @@
+use {
+    crate::{
+        configs::evm_config::EvmChainName,
+        types::{
+            hyperlane::{CheckpointMatchEvent, DispatchUpdate},
+            pragma::{
+                calldata::{AsCalldata, HyperlaneMessage, Payload},
+                constants::{DEFAULT_ACTIVE_CHAIN, HYPERLANE_VERSION, MAX_CLIENT_MESSAGE_SIZE, PING_INTERVAL_DURATION},
+            },
+            rpc::RpcDataFeed,
+        },
+        AppState,
+    },
+    alloy::hex,
+    anyhow::{anyhow, Result},
+    axum::{
+        extract::{
+            ws::{Message, WebSocket, WebSocketUpgrade},
+            ConnectInfo, State as AxumState,
+        },
+        response::IntoResponse,
+    },
+    futures::{
+        stream::{SplitSink, SplitStream},
+        SinkExt, StreamExt,
+    },
+    serde::{Deserialize, Serialize},
+    std::{
+        collections::HashMap,
+        net::SocketAddr,
+        sync::{
+            atomic::{AtomicUsize, Ordering},
+            Arc,
+        },
+    },
+    tokio::sync::{broadcast::Receiver, watch},
+};
+
+#[derive(Clone)]
+pub struct DataFeedClientConfig {}
+
+pub struct WsState {
```

Comment:
We should define this `WsState` somewhere else - it looks like this is specific to this endpoint. Should probably live close to `AppState`?